storage/namespace.go

```diff
@@ -0,0 +1,280 @@
+	shardID := n.shardSet.Shard(id)
+	shard, err := n.shardAt(shardID)
+	if err != nil {
+		return nil, err
+	}
+	return shard.ReadEncoded(ctx, id, start, end)
+}
+
+func (n *dbNamespace) FetchBlocks(
+	ctx context.Context,
+	shardID uint32,
+	id string,
+	starts []time.Time,
+) ([]FetchBlockResult, error) {
+	shard, err := n.shardAt(shardID)
+	if err != nil {
+		return nil, xerrors.NewInvalidParamsError(err)
+	}
+	return shard.FetchBlocks(ctx, id, starts), nil
+}
+
+func (n *dbNamespace) FetchBlocksMetadata(
+	ctx context.Context,
+	shardID uint32,
+	limit int64,
+	pageToken int64,
+	includeSizes bool,
+) ([]FetchBlocksMetadataResult, *int64, error) {
+	shard, err := n.shardAt(shardID)
+	if err != nil {
+		return nil, nil, xerrors.NewInvalidParamsError(err)
+	}
+	res, nextPageToken := shard.FetchBlocksMetadata(ctx, limit, pageToken, includeSizes)
+	return res, nextPageToken, nil
+}
+
+func (n *dbNamespace) Bootstrap(
+	bs bootstrap.Bootstrap,
+	writeStart time.Time,
+	cutover time.Time,
+) error {
+	n.Lock()
+	if n.bs == bootstrapped {
+		n.Unlock()
+		return nil
+	}
+	if n.bs == bootstrapping {
+		n.Unlock()
+		return errNamespaceIsBootstrapping
+	}
+	n.bs = bootstrapping
+	n.Unlock()
+
+	if !n.nopts.GetNeedsBootstrap() {
+		n.Lock()
+		n.bs = bootstrapped
+		n.Unlock()
+		return nil
+	}
+
+	multiErr := xerrors.NewMultiError()
+	shards := n.getOwnedShards()
+	for _, shard := range shards {
+		err := shard.Bootstrap(bs, n.name, writeStart, cutover)
+		multiErr = multiErr.Add(err)
+	}
+
+	n.Lock()
+	n.bs = bootstrapped
+	n.Unlock()
+
+	return multiErr.FinalError()
+}
+
+func (n *dbNamespace) Flush(
+	ctx context.Context,
+	blockStart time.Time,
+	pm persist.Manager,
+) error {
+	n.RLock()
+	if n.bs != bootstrapped {
+		n.RUnlock()
+		return errNamespaceNotBootstrapped
+	}
+	n.RUnlock()
+
+	if !n.nopts.GetNeedsFlush() {
+		return nil
+	}
+
+	multiErr := xerrors.NewMultiError()
+	shards := n.getOwnedShards()
+	for _, shard := range shards {
+		// NB(xichen): we still want to proceed if a shard fails to flush its data.
+		// Probably want to emit a counter here, but for now just log it.
+		if err := shard.Flush(ctx, n.name, blockStart, pm); err != nil {
+			detailedErr := fmt.Errorf("shard %d failed to flush data: %v", shard.ID(), err)
+			multiErr = multiErr.Add(detailedErr)
+		}
+	}
+
+	return multiErr.FinalError()
+}
+
+func (n *dbNamespace) CleanupFileset(earliestToRetain time.Time) error {
+	if !n.nopts.GetNeedsFilesetCleanup() {
+		return nil
+	}
+
+	multiErr := xerrors.NewMultiError()
+	shards := n.getOwnedShards()
+	for _, shard := range shards {
+		if err := shard.CleanupFileset(n.name, earliestToRetain); err != nil {
+			multiErr = multiErr.Add(err)
+		}
+	}
+
+	return multiErr.FinalError()
+}
+
+func (n *dbNamespace) Truncate() (int64, error) {
+	var totalNumSeries int64
+
+	n.RLock()
+	shards := n.shardSet.Shards()
+	for _, shard := range shards {
+		totalNumSeries += n.shards[shard].NumSeries()
+	}
+	n.RUnlock()
+
+	n.initShards()
```

Comment:
I was hoping GC will kick in and reclaim the memory after the shards are recreated. After all nothing is holding onto them anyways. I'd imagine if we ever truncate a namespace, it's likely for the purpose of dropping memory footprint, in which case it'll be done when GC kicks in.